### PR TITLE
WIP: Monitoring of non-Theano quantities on training data.

### PR DESCRIPTION
**Pull request by rizar**
For #1006.

This is the very first code that worked for me.

### CODE REVIEW

File: blocks/extensions/monitoring.py [hunk 1]
```diff
@@ -122,22 +134,39 @@ def do(self, callback_name, *args):
         aggregation buffer and instructs the training algorithm what
         additional computations should be carried at each step by adding
         corresponding updates to it. In all other cases it writes
-        aggregated values of the monitored variables to the log.
+aggregated values of the monitored variables to the log.
```
Comment (dmitriy-serdyuk) [inline blocks/extensions/monitoring.py]: Something wrong with indention here.

File: blocks/extensions/monitoring.py [hunk 2]
```diff
@@ -122,22 +134,39 @@ def do(self, callback_name, *args):
...
         if callback_name == 'before_training':
             if not isinstance(self.main_loop.algorithm,
                               DifferentiableCostMinimizer):
                 raise ValueError
             self.main_loop.algorithm.add_updates(
-                self._buffer.accumulation_updates)
-            self._buffer.initialize_aggregators()
+                self._variables.accumulation_updates)
+            self.main_loop.algorithm.add_updates(
+                self._required.accumulation_updates)
```
Comment (dmitriy-serdyuk) [inline blocks/extensions/monitoring.py]: Shouldn't you add updates for required variables before the variables?

Comment (rizar) [inline blocks/extensions/monitoring.py]: I don't think that the order matters here. These updates will act on different shared variables.

**Comment by rizar** [comment]
By the way, method names in classes related to monitoring are so random:

```python
class MonitoredQuantity(object):
     def initialize()
     def accumulate()
     def readout()

class MonitoredQuantityBuffer(object):
     def initialize()
     def accumulate_quantities()
     def get_aggregated_values()

class AggregationBuffer(object):
    def initialize_aggregators()
    def get_aggregated_values()
```

I guess a cleanup is needed, e.g. instead of `accumulate` we can always use `aggregate`. `initialize_aggregators` can be just `initialize`, `readout` can become `get_aggregated_value`. Opinions?

**Comment by rizar** [comment]
parity_problem is broken because I renamed the input of `TrainingDataMonitoring` from `variables` to `observables`.  Not sure if that was a good decision, because after all, usually it is variables that is monitored. Opinions?

**Comment by dwf** [comment]
Observables is a weird choice. Sounds like some sort of graphical model
thing. I'd stick with variables.

**Comment by rizar** [comment]
Variables it is. Tests pass. What about the renaming that I proposed above?

**Comment by rizar** [comment]
Assuming that the tests will pass, this is now ready for a final review.

**Comment by dmitriy-serdyuk** [comment]
This looks good to me.

A note for future: `requires` syntax (like [here](https://github.com/mila-udem/blocks/pull/1007/files#diff-9abcd6167516efa783ed70442ad2c666R56)) looks a little bit weird. And the signature of `aggregate` of the abstract `MonitoredQuantity` doesn't suppose to input any additional parameters. Maybe we should refactor it to be able to input a dictionary `{name: variable}` and pass it by name.

**Comment by rizar** [comment]
I think that "requires" argument currently is pretty convenient. The signature of `aggregate` is actually quite horrible, I will improve it a bit before the merge.

**Pull request merged**